lua/nvim-treesitter/locals.lua

```diff
@@ -0,0 +1,80 @@
+-- Functions to handle locals
+-- Locals are a generalization of definition and scopes
+-- its the way nvim-treesitter uses to "understand" the code
+local api = vim.api
+local ts = vim.treesitter
+local queries = require'nvim-treesitter.query'
+local parsers = require'nvim-treesitter.parsers'
+
+local M = {
+  locals={}
+}
+
+function M.is_supported(lang)
+  return queries.get_query(lang, "locals") ~= nil
+end
+
+function M.collect_locals(bufnr)
+  local ft = api.nvim_buf_get_option(bufnr, "ft")
+  local query = queries.get_query(ft, "locals")
+
+  local parser = parsers.get_parser(bufnr)
+  local root = parser:parse():root()
+  local start_row, _, end_row, _ = root:range()
+
+  local locals = {}
+
+  for prepared_match in queries.iter_prepared_matches(query, root, bufnr, start_row, end_row) do
+    table.insert(locals, prepared_match)
+  end
+
+  return locals
+end
+
+function M.on_lines(_, buf, _, firstline, lastline, new_lastline, _)
+  M.locals[buf] = M.collect_locals(buf)
+end
+
+function M.get_definitions(bufnr)
+  local locals = M.locals[bufnr or api.nvim_get_current_buf()]
+
+  local defs = {}
+
+  for _, loc in ipairs(locals) do
+    if loc.definition then
+      table.insert(defs, {definition=loc.definition, kind=loc.kind})
+    end
+  end
+
+  return defs
+end
+
+function M.get_scopes(bufnr)
+  local locals = M.locals[bufnr or api.nvim_get_current_buf()]
+
+  local scopes = {}
+
+  for _, loc in ipairs(locals) do
+    if loc.scope then
+      table.insert(scopes, loc.scope)
+    end
+  end
+
+  return scopes
+end
+
+function M.get_references(bufnr)
+  local locals = M.locals[bufnr or api.nvim_get_current_buf()]
+
+  local refs = {}
+
+  for _, loc in ipairs(locals) do
```

Comment:
also this is more about naming conventions, but you could call your local `local` and not `loc`. It doesn't matters much though, loc is semi explicit already.

Reply:
hum my bad it's a reserved word in lua

Reply:
Yeah that's why I called it this way but a better name is effectively needed.